src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/CalculationTransactionalService.kt

```diff
@@ -39,18 +42,117 @@ class CalculationTransactionalService(
+   * 2. Validate the transformation of the raw Booking (NOMIS offence data transformed into a Booking object)
+   * 3. Run the calculation and catch any errors thrown by the calculation algorithm
+   * 4. Validate the post calculation Booking (The Booking is transformed during the calculation). e.g. Consecutive sentences (aggregates)
+   */
+  fun fullValidation(prisonerId: String, calculationUserInputs: CalculationUserInputs): List<ValidationMessage> {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId)
+    var messages = validationService.validateSourceData(sourceData) // Validation stage 1 of 4
+    if (messages.isNotEmpty()) return messages
+
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    messages = validationService.validateBookingBeforeCalculation(booking) // Validation stage 2 of 4
+    if (messages.isNotEmpty()) return messages
+    try {
+      val bookingAfterCalculation = calculationService.calculate(booking) // Validation stage 3 of 4
+      validationService.validateBookingAfterCalculation(bookingAfterCalculation) // Validation stage 4 of 4
+    } catch (validationException: CrdCalculationValidationException) {
+      return listOf(ValidationMessage(validationException.validation))
+    }
+    return messages
+  }
+
+  //  The activeDataOnly flag is only used by a test endpoint (1000 calcs test, which is used to test historic data)
+  @Transactional
+  fun calculate(prisonerId: String, calculationUserInputs: CalculationUserInputs, activeDataOnly: Boolean = true): CalculatedReleaseDates {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId, activeDataOnly)
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    try {
+      return calculate(booking, PRELIMINARY, sourceData, calculationUserInputs)
+    } catch (error: Exception) {
+      recordError(booking, sourceData, calculationUserInputs, error)
+      throw error
+    }
+  }
+
+  @Transactional
+  fun validateAndConfirmCalculation(
+    calculationRequestId: Long,
+    calculationFragments: CalculationFragments,
+  ): CalculatedReleaseDates {
+    val calculationRequest =
+      calculationRequestRepository.findByIdAndCalculationStatus(
+        calculationRequestId,
+        PRELIMINARY.name
+      ).orElseThrow {
+        EntityNotFoundException("No preliminary calculation exists for calculationRequestId $calculationRequestId")
+      }
+    val sourceData = prisonService.getPrisonApiSourceData(calculationRequest.prisonerId)
+    val userInput = transform(calculationRequest.calculationRequestUserInput)
+    val booking = bookingService.getBooking(sourceData, userInput)
+
+    if (calculationRequest.inputData.hashCode() != objectToJson(booking, objectMapper).hashCode()) {
+      throw PreconditionFailedException("The booking data used for the preliminary calculation has changed")
+    }
+
+    if (validationService.validateSourceData(sourceData).isNotEmpty()) {
```

Comment:
I think validation should be run before the booking is created.

Reply:
agreed, I think it would make sense. To discuss.  It'a actually the same order as before

Reply:
Potentially, remove validation - deal with in next PR (discuss with Joel too)